common/network-common/src/main/java/org/apache/spark/network/crypto/GcmTransportCipher.java

```diff
@@ -0,0 +1,356 @@
+            this.encryptedCount =
+                    LENGTH_HEADER_BYTES + aesGcmHkdfStreaming.expectedCiphertextSize(bytesToRead);
+        }
+
+        @Override
+        public long position() {
+            return 0;
+        }
+
+        @Override
+        public long transferred() {
+            return transferred;
+        }
+
+        @Override
+        public long count() {
+            return encryptedCount;
+        }
+
+        @Override
+        public long transferTo(WritableByteChannel target, long position) throws IOException {
+            Preconditions.checkArgument(position == transferred(),
+                    "Invalid position.");
+            int transferredThisCall = 0;
+            // The format of the output is:
+            // [8 byte length][Internal IV and header][Ciphertext][Auth Tag]
+            if (!headerWritten) {
+                ByteBuffer expectedLength = ByteBuffer
+                        .allocate(LENGTH_HEADER_BYTES)
+                        .putLong(encryptedCount)
+                        .flip();
+                target.write(expectedLength);
+                int headerWritten = LENGTH_HEADER_BYTES + target.write(encrypter.getHeader());
+                transferredThisCall += headerWritten;
+                this.transferred += headerWritten;
+                this.headerWritten = true;
+            }
+
+            while (bytesRead < bytesToRead) {
+                long readableBytes = getReadableBytes();
+                boolean lastSegment = readableBytes <= plaintextBuffer.capacity();
+                plaintextBuffer.clear();
+                int readLimit =
+                        (int) Math.min(readableBytes, plaintextBuffer.capacity());
+                plaintextBuffer.limit(readLimit);
+                if (plaintextMessage instanceof ByteBuf byteBuf) {
+                    byteBuf.readBytes(plaintextBuffer);
+                    long inputBytesRead = readableBytes - byteBuf.readableBytes();
+                    bytesRead += inputBytesRead;
+                } else if (plaintextMessage instanceof AbstractFileRegion fileRegion) {
+                    ByteBufferWriteableChannel plaintextChannel =
+                            new ByteBufferWriteableChannel(plaintextBuffer);
+                    long transferred =
+                            fileRegion.transferTo(plaintextChannel, fileRegion.transferred());
+                    bytesRead += transferred;
+                    if (transferred == 0) {
+                        // File regions may return 0 if they are not ready to transfer
+                        // more data. In that case, we'll return with the expectation
+                        // that this transferTo() is called again.
+                        return transferredThisCall;
+                    }
+                }
+                plaintextBuffer.flip();
+                ciphertextBuffer.clear();
```

Comment:
QQ: If `plaintextBuffer.remaining() != plaintextMessage.capacity()`, the segment encryption will end up with incorrect sizes, no ?
For `ByteBuf` case this should not happen, but can for `FileRegion` reads.

Reply:
Yes, I think in the current implementation this would break. The decryption is written assuming that segments will be full. (It could conceivably handle variable length segments, but then we'd need to pass per-segment lengths as well.)

Would it make sense to change the check in the FileRegion case to be:
```
if (transferred < readLimit) {
  return transferredThisCall;
}
```

If we don't read the limit of what we should be able to, then don't encrypt this block and wait until `writeTo` is called again to flush it out.

Reply:
Yes, we will have to wait for `plaintextBuffer` to fill up before encrypting.
If it was unable to fill, return `transferredThisCall` and wait for a retry.